plugins/woocommerce/client/blocks/assets/js/base/context/hooks/cart/use-store-cart.ts

```diff
@@ -122,122 +121,113 @@ export const defaultCartData: StoreCart = {
-			const { receiveCart, receiveCartContents } = dispatch( cartStore );
-
-			const cartFees =
-				cartData.fees.length > 0
-					? cartData.fees.map( ( fee: CartResponseFeeItem ) =>
-							decodeValues( fee )
-					  )
-					: EMPTY_CART_FEES;
-
-			// Add a text property to the coupon to allow extensions to modify
-			// the text used to display the coupon, without affecting the
-			// functionality when it comes to removing the coupon.
-			const cartCoupons: CartResponseCoupons =
-				cartData.coupons.length > 0
-					? cartData.coupons.map(
-							( coupon: CartResponseCouponItem ) => ( {
-								...coupon,
-								label: coupon.code,
-							} )
-					  )
-					: EMPTY_CART_COUPONS;
-
-			// Update refs to keep the hook stable.
-			const billingAddress = emptyHiddenAddressFields(
-				decodeValues( cartData.billingAddress )
-			);
-			const shippingAddress = cartData.needsShipping
-				? emptyHiddenAddressFields(
-						decodeValues( cartData.shippingAddress )
-				  )
-				: billingAddress;
-
-			if (
-				! fastDeepEqual( billingAddress, billingAddressRef.current )
-			) {
-				billingAddressRef.current = billingAddress;
-			}
-
-			if (
-				! fastDeepEqual( shippingAddress, shippingAddressRef.current )
-			) {
-				shippingAddressRef.current = shippingAddress;
-			}
-
-			return {
-				cartCoupons,
-				cartItems: cartData.items,
-				crossSellsProducts: cartData.crossSells,
-				cartFees,
-				cartItemsCount: cartData.itemsCount,
-				cartItemsWeight: cartData.itemsWeight,
-				cartNeedsPayment: cartData.needsPayment,
-				cartNeedsShipping: cartData.needsShipping,
-				cartItemErrors: cartData.errors,
-				cartTotals,
-				cartIsLoading,
-				cartErrors,
-				billingData: billingAddressRef.current,
-				billingAddress: billingAddressRef.current,
-				shippingAddress: shippingAddressRef.current,
-				extensions: cartData.extensions,
-				shippingRates: cartData.shippingRates,
-				isLoadingRates,
-				cartHasCalculatedShipping: cartData.hasCalculatedShipping,
-				paymentRequirements: cartData.paymentRequirements,
-				receiveCart,
-				receiveCartContents,
-			};
-		},
-		[ shouldSelect ]
+	const { receiveCart, receiveCartContents } = useDispatch( cartStore );
+	const { cartData, cartErrors, cartTotals, cartIsLoading, isLoadingRates } =
+		useSelect(
+			( select ) => {
+				const store = select( cartStore );
+				return {
+					cartData: store.getCartData(),
+					cartErrors: store.getCartErrors(),
+					cartTotals: store.getCartTotals(),
+					cartIsLoading:
+						// @ts-expect-error `hasFinishedResolution` is not typed in @wordpress/data yet.
+						! store.hasFinishedResolution( 'getCartData' ),
+					isLoadingRates: store.isCustomerDataUpdating(),
+				};
+			},
+			[ shouldSelect ]
+		);
+
+	const cartFees = useMemo( () => {
+		return cartData.fees.length > 0
+			? cartData.fees.map( ( fee: CartResponseFeeItem ) =>
+					decodeValues( fee )
+			  )
+			: EMPTY_CART_FEES;
+	}, [ JSON.stringify( cartData.fees ) ] );
```

Comment:
You shouldn't do this,  `JSON.stringify` is expensive and would run on each render, negates the point of `useMemo`.

Also `JSON.stringify` creates a new string on each render, then `useMemo` compares it to the old one, it won't match because they are _referentially_ not equal, the memoized function will therefore recompute.

I think just using `cartData.fees` might be OK here, I think it is the same object unless the state is changed in the data store, so if this hook reruns but data store hasn't changed it would not recompute.

Reply:
Hmm I thought we'd used this method before but its not in the codebase. https://medium.com/@ivanreshetar/enhanced-strategies-for-using-objects-and-arrays-as-dependencies-in-react-hooks-43bbfa4f20d2

`cartData.fees` is an object so I think it will have issues. I'll have to test and see if its running on every render.

Reply:
https://github.com/woocommerce/woocommerce/pull/57673/commits/a25d487e8089f13b7ab1da1d9e29df2836967700